exquisite_corpus/spm_naive_language_model.py

```diff
@@ -0,0 +1,428 @@
+        n_directions = 2 if self.rnn.bidirectional else 1
+        full_num_layers = self.rnn.num_layers * n_directions
+        if hidden_states is None:
+            hidden_states = torch.zeros(
+                full_num_layers, batch_size, self.rnn.hidden_size
+            ).to(input.device, non_blocking=True)
+        output = input
+        output = self.encoder(output)
+        output, new_hidden_states = self.rnn(
+            output,
+            hidden_states.view(full_num_layers, batch_size, self.rnn.hidden_size),
+        )
+        output = self.decoder(output)
+        output = self.log_softmax(output)
+        if return_hidden:
+            new_hidden_states = new_hidden_states.view(hidden_states.size())
+            return output, new_hidden_states
+        else:
+            return output
+
+
+class LossFunction(nn.Module):
+    """
+    Module to compute the aggregate (over all elements of a batch, and all
+    sequence positions within each element) loss.
+    """
+
+    def __init__(self, reduction="mean"):
+        super().__init__()
+        self.loss_fn = nn.NLLLoss(reduction=reduction)
+
+    def forward(self, input, target):
+        return self.loss_fn(input.view(target.numel(), -1), target.view(-1))
+
+
+class ModelManager:
+    """
+    Objects that own torch models (modules) and provide convenience methods
+    for their training and use.
+    """
+
+    def __init__(
+        self,
+        model=None,
+        device=torch.device("cuda:0"),
+        **kwargs
+    ):
+        if model is None:
+            self.model = LanguageModel(**kwargs)
+        else:
+            self.model = model
+        self.device = device
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, new_device):
+        if isinstance(new_device, str):
+            new_device = torch.device(new_device)
+        if new_device != torch.device("cpu") and not torch.cuda.is_available():
+            print("Warning:  cuda not available so setting device to CPU.")
+            self._device = torch.device("cpu")
+        else:
+            self._device = new_device
+        self.model.to(self._device)
+
+    def collate_batch_with_labels(self, raw_batch):
+        """
+        Take a raw batch (as returned by a DataLoader and SpmIdsBatchSampler),
+        collate it, and split off labels.
+        """
+        data = torch.LongTensor(np.vstack([x[:-1] for x in raw_batch]))
+        labels = torch.LongTensor(np.vstack([x[1:] for x in raw_batch]))
+        return data, labels
+
+    def train(
+        self,
+        train_dataset_path,
+        validation_dataset_path=None,
+        save_path=None,
+        min_length=2,
+        n_epochs=np.inf,
+        n_batches_between_saves=1000,
+        n_batches_between_messages=50,
+        n_batches_between_validations=25000,
+        batch_size=256,
+        n_data_loader_workers=0,
+        lr=0.05,
+        device=None,
+    ):
```

Comment:
I understand that there's a lot that needs to happen in training, but this method is long and confusing and contains pieces whose purpose I don't understand. It also surprises me that we have to spell out so much about how to train a language model.

So:
- The method should have a docstring.
- If all this code is necessary to be in the train method, it should at least have more frequent explanatory comments.
- Why is this code that we had to write?

Reply:
I've added docstrings and documentation as requested.  There is code to produce progress messages and save and restore checkpoints during training, which aren't strictly necessary but seem pretty desirable to me.  Was there something specific you think could be cut?